src/dag.c

```diff
@@ -231,52 +332,221 @@ void RedisAI_DagRunSession_ScriptRun_Step(RedisAI_RunInfo *rinfo, RAI_DagOp *cur
-  int device_complete_ = 1;
+  RAI_DagOp *currentOp_ = NULL;
   for (size_t i = 0; i < array_len(rinfo->dagOps); i++) {
 
     if (rinfo->dagOps[i]->result >= 0) {
       continue;
     }
 
-    all_devices_complete_ = 0;
-
     if (strcasecmp(devicestr, rinfo->dagOps[i]->devicestr) == 0) {
-      device_complete_ = 0;
-      currentOp = rinfo->dagOps[i];
+      currentOp_ = rinfo->dagOps[i];
       break;
     }
   }
 
-  *device_complete = device_complete_;
-  *all_devices_complete = all_devices_complete_;
+  *currentOp = currentOp_;
+
+  *rinfo->dagRefCount -= 1;
+}
  
-  if (currentOp == NULL && device_complete) {
+void RedisAI_DagCurrentOpAndInfo(RedisAI_RunInfo *rinfo, const char *devicestr,
+                                 RAI_DagOp **currentOp, int *currentOpReady,
+                                 int *currentOpBatchable,
+                                 int *deviceComplete, int *dagComplete) {
+  pthread_mutex_lock(rinfo->dagMutex);
+
+  *rinfo->dagRefCount += 1;
+
+  RAI_DagOp *currentOp_ = NULL;
+  int dagComplete_ = 1;
+  int deviceComplete_ = 1;
+  for (size_t i = 0; i < array_len(rinfo->dagOps); i++) {
+    if (rinfo->dagOps[i]->result >= 0) {
+      continue;
+    }
+
+    dagComplete_ = 0;
+
+    if (strcasecmp(devicestr, rinfo->dagOps[i]->devicestr) == 0) {
+      deviceComplete_ = 0;
+      currentOp_ = rinfo->dagOps[i];
+      break;
+    }
+  }
+
+  *currentOp = currentOp_;
+  *deviceComplete = deviceComplete_;
+  *dagComplete = dagComplete_;
+  *currentOpReady = 0;
+  *currentOpBatchable = 0;
+
+  if (deviceComplete_) {
     *rinfo->dagRefCount -= 1;
     pthread_mutex_unlock(rinfo->dagMutex);
     return;
   }
 
-  uint n_inkeys = array_len(currentOp->inkeys);
+  if (currentOp_->mctx && currentOp_->mctx->model->opts.batchsize > 0) {
+    *currentOpBatchable = 1;
+  }
+
+  uint n_inkeys = array_len(currentOp_->inkeys);
+  *currentOpReady = 1;
   for (int i=0; i<n_inkeys; i++) {
     if (AI_dictFind(rinfo->dagTensorsContext,
-                    RedisModule_StringPtrLen(currentOp->inkeys[i], NULL)) == NULL) {
+                    RedisModule_StringPtrLen(currentOp_->inkeys[i], NULL)) == NULL) {
       *rinfo->dagRefCount -= 1;
       pthread_mutex_unlock(rinfo->dagMutex);
-      *progress = 0;
+      *currentOpReady = 0;
       return;
     }
   }
 
+  *rinfo->dagRefCount -= 1;
+
+  pthread_mutex_unlock(rinfo->dagMutex);
+}
+
+void RedisAI_DagOpBatchInfo(RedisAI_RunInfo *rinfo, RAI_DagOp *op,
+                            size_t *batchsize, size_t *minbatchsize, size_t *inbatchsize) {
+  *batchsize = 0;
+  *minbatchsize = 0;
+  *inbatchsize = 0;
+ 
+  pthread_mutex_lock(rinfo->dagMutex);
+
+  *rinfo->dagRefCount += 1;
+
+  if (op->mctx) {
+    *batchsize = op->mctx->model->opts.batchsize;
+    *minbatchsize = op->mctx->model->opts.minbatchsize;
+    *inbatchsize = RAI_DagOpBatchSize(op, rinfo->dagTensorsContext);
+  }
+
+  *rinfo->dagRefCount -= 1;
+
+  pthread_mutex_unlock(rinfo->dagMutex);
```

Comment:
why locking the dag here?

Reply:
Because `RAI_DagOpBatchSize` needs to reach into the local context.